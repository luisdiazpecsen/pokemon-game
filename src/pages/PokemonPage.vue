<template>
    <h1 v-if="!pokemon">Please, wait...</h1>
    <div v-else>
        <!-- Pokemon -->
        <h1>Who's this Pok√©mon?</h1>
        <PokemonPicture
            v-bind:pokemon-id="pokemon.id"
            :show-pokemon="showPokemon"
        />

        <!-- Options -->
        <PokemonOptions
            v-if="!showAnswer"
            :pokemons="pokemonArr"
            @select-pokemon="checkAnswer(1, $event)"
        />

        <template v-else>
            <h2 class="fade-in">{{ message }}</h2>
            <button @click="newGame" class="fade-in">New game</button>
        </template>
    </div>
</template>

<script>
// Components
import PokemonPicture from "@/components/PokemonPicture.vue"
import PokemonOptions from "@/components/PokemonOptions.vue"

// Helpers
import getPokemonOptions from "@/helpers/getPokemonOptions"

const ANSWERS = {
    ok: "THAT'S RIGHT!",
    wrong: "OOPS!",
}

export default {
    name: "PokemonPage",
    components: {
        PokemonPicture,
        PokemonOptions,
    },
    data: function () {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: "",
        }
    },
    methods: {
        mixPokemonArray: async function () {
            this.pokemonArr = await getPokemonOptions()
            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[rndInt]
        },
        checkAnswer: function (test, selectedId) {
            this.showPokemon = true

            if (selectedId === this.pokemon.id) {
                this.message = `${ANSWERS.ok} - It's ${this.pokemon.name}`
            } else {
                this.message = `${ANSWERS.wrong} - It's ${this.pokemon.name}`
            }

            this.showAnswer = true
        },
        newGame: function () {
            this.pokemonArr = []
            this.pokemon = null
            this.showPokemon = false
            this.showAnswer = false
            this.message = ""
            this.mixPokemonArray()
        },
    },
    mounted: function () {
        this.mixPokemonArray()
    },
}
</script>