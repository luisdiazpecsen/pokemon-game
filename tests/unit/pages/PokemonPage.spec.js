import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from "@/pages/PokemonPage"
import { pokemons } from '../mocks/pokemons.mock'
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

describe('PokemonPage component', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('should call mixPokemonArray after mount', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('should match with snapshot when pokémons are loaded', () => {
        const wrapper = shallowMount(PokemonPage, {
            data: function () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('should show PokemonPicture and PokemonOptions components', () => {
        /**
         * PokemonPicture should exist
         * PokemonOptions should exist
         * 
         * PokemonPicture should have pokemonId attribute and should be equal to the first pokémon
         * PokemonOptions should have pokemons attribute (to be truthy)
         */
        const wrapper = mount(PokemonPage, {
            data: function () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        // const pokemonPicture = wrapper.find('pokemon-picture-stub')
        // expect(pokemonPicture.exists()).toBeTruthy()

        // const pokemonOptions = wrapper.find('pokemon-options-stub')
        // expect(pokemonOptions.exists()).toBeTruthy()

        // expect(pokemonPicture.attributes('pokemonid')).toBe(pokemons[0].id.toString())
        // expect(pokemonOptions.attributes('pokemons')).toBeTruthy()

        const pokemonPicture = wrapper.findComponent(PokemonPicture)
        expect(pokemonPicture.exists()).toBeTruthy()

        const pokemonOptions = wrapper.findComponent(PokemonOptions)
        expect(pokemonOptions.exists()).toBeTruthy()

        const { pokemonId } = pokemonPicture.props()
        expect(pokemonId).toBe(pokemons[0].id)

        const { pokemons: options } = pokemonOptions.props()
        expect(options).toStrictEqual(pokemons)
    })

    test('testing checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data: function () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        })

        const ANSWERS = {
            ok: "THAT'S RIGHT!",
            wrong: "OOPS!",
        }

        await wrapper.vm.checkAnswer(1, pokemons[0].id)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`${ANSWERS.ok} - It's ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(1, pokemons[1].id)
        expect(wrapper.vm.message).toBe(`${ANSWERS.wrong} - It's ${pokemons[0].name}`)
    })
})