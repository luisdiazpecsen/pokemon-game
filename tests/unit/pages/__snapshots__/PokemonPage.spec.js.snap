// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match with snapshot 1`] = `<h1>Please, wait...</h1>`;

exports[`PokemonPage component should match with snapshot when pokémons are loaded 1`] = `
<div>
  <!-- Pokemon -->
  <h1>Who's this Pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- Options -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
</div>
`;
